### PR TITLE
Fix Coverity 1503096: out-of-bounds access

### PR DESCRIPTION
**Pull request by paulidale**
Coverity is claiming that `n` can be 48 which is larger than the IV size.  It's not clear how/why this is possible but other paths do check for a maximum IV length, so adding it here seems reasonable.

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/evp/evp_enc.c
Comment (t8m) [inline crypto/evp/evp_enc.c:369]: Previously we did not assert on n <= 0. But OK

**Review by t8m — APPROVED** [review]

File: crypto/evp/evp_enc.c
Comment (paulidale) [reply crypto/evp/evp_enc.c:369]: Earlier in the switch statement, there is a similar assert.  I figure it was appropriate here.

**Comment by paulidale** [comment]
Ping @openssl/committers for second review

**Comment by paulidale** [comment]
Reping https://github.com/orgs/openssl/teams/committers for second review

File: crypto/evp/evp_enc.c
Comment (mspncp) [reply crypto/evp/evp_enc.c:369]: Do you mean the assert in the `EVP_CIPH_CBC_MODE`? It was added a year ago (by you) in commit c12bf35026af94a73402eaf13f2428a9af30f1c0 to fix another coverity warning.

**Review by mspncp — COMMENTED** [review]
Searching the source code (`git grep -n -A 10  EVP_CIPHER_CTX_get_iv_length`) it seems like we are not very consistent about checking for a negative return value (which indicates an error). In many cases we push a runtime error on the stack, which seems to me to be more appropriate than asserting.

https://github.com/openssl/openssl/blob/ea2f9006c3d86845e391f73ff9eccf0c51efd44c/crypto/evp/e_aes.c#L1124-L1134


In other locations, the function returns withiout pushing an error onto the stack

https://github.com/openssl/openssl/blob/ea2f9006c3d86845e391f73ff9eccf0c51efd44c/crypto/evp/e_aes.c#L3699-L3700

even though the analogue error for the key lenght is reported.

https://github.com/openssl/openssl/blob/ea2f9006c3d86845e391f73ff9eccf0c51efd44c/crypto/evp/e_aes.c#L3687-L3690

Also, we are not consistent in whether we check for `<= 0` or `< 0`. Maybe it would be a good idea to check all those locations and make the error handling more consistent?

**Comment by mspncp** [comment]
> Also, we are not consistent in whether we check for <= 0 or < 0.

Does it actually make sense to allow an IV size which is smaller than the block size of the cipher, or shouldn't this be considered an error as well?

**Comment by beldmit** [comment]
> Does it actually make sense to allow an IV size which is smaller than the block size of the cipher, or shouldn't this be considered an error as well?

It is (sort of) valid for Russian GOST counter modes where we have composite IV having length half block length and counter as a 2nd half. If we consider that length of IV is equal to block size, we will have problems (and hard-coding) generating IV in TLS.

**Comment by paulidale** [comment]
I'm checking for overflowing the structure's array which would very bad.

Checking for negative returns etc is another matter -- my guess is most have been flagged as false positives and that they will reappear eventually.  We should consistently raise an error in these cases, but not in this PR.

File: crypto/evp/evp_enc.c
Comment (paulidale) [reply crypto/evp/evp_enc.c:369]: It was an assert already, I just improved it :)

**Comment by mspncp** [comment]
> > Does it actually make sense to allow an IV size which is smaller than the block size of the cipher, or shouldn't this be considered an error as well?
> 
> It is (sort of) valid for Russian GOST counter modes where we have composite IV having length half block length and counter as a 2nd half. If we consider that length of IV is equal to block size, we will have problems (and hard-coding) generating IV in TLS.

It is not uncommon that the IV is composed of a shorter IV and a counter. But wouldn't the length of the stored IV effectively be the length of the composed one, not of the shorter half?

File: crypto/evp/evp_enc.c
Comment (mspncp) [reply crypto/evp/evp_enc.c:369]: Ok, so this particular location is not your fault ;-). Nevertheless, we're inconsistent.

**Comment by beldmit** [comment]
https://www.ietf.org/rfc/rfc9189.html#name-ctr_omac

File: crypto/evp/evp_enc.c
Comment (mattcaswell) [reply crypto/evp/evp_enc.c:369]: It is actually possible that `EVP_CIPHER_CTX_get_iv_length` may return -1. So an assert does not seem appropriate.

Comment (mspncp) [reply crypto/evp/evp_enc.c:369]: That was exactly my point in [my review](https://github.com/openssl/openssl/pull/17898#pullrequestreview-922769641).

**Comment by paulidale** [comment]
Changed to returning an error on negative.

BTW: ossl_assert only asserts in debug builds which is less harmful.

**Comment by mspncp** [comment]
> BTW: ossl_assert only asserts in debug builds which is less harmful.

Agreed, but the essential point is that it does not add `EVP_R_INVALID_IV_LENGTH` to the error stack, as we do elsewhere.

**Comment by paulidale** [comment]
Added errors to the two cases touched here.

File: crypto/evp/evp_enc.c
Comment (mspncp) [inline crypto/evp/evp_enc.c:358]:
```suggestion
                ERR_raise(ERR_LIB_EVP, EVP_R_INVALID_IV_LENGTH);
```

File: crypto/evp/evp_enc.c
Comment (mspncp) [inline crypto/evp/evp_enc.c:372]:
```suggestion
                    ERR_raise(ERR_LIB_EVP, EVP_R_INVALID_IV_LENGTH);
```

**Review by mspncp — CHANGES_REQUESTED** [review]

File: crypto/evp/evp_enc.c
Comment (paulidale) [reply crypto/evp/evp_enc.c:372]: Fixed both, it's been a long day.

File: crypto/evp/evp_enc.c
Comment (t8m) [reply crypto/evp/evp_enc.c:358]: This does not seem to be fixed yet

File: crypto/evp/evp_enc.c
Comment (t8m) [reply crypto/evp/evp_enc.c:372]: You probably did not push.

Comment (mspncp) [reply crypto/evp/evp_enc.c:372]: It must have been a _really long_ day.

Comment (paulidale) [reply crypto/evp/evp_enc.c]: Yeah, forgot to push.

File: crypto/evp/evp_enc.c
Comment (t8m) [inline crypto/evp/evp_enc.c:357]: The code allowed 0 length iv before. Not sure if we should change it.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/evp/evp_enc.c
Comment (paulidale) [reply crypto/evp/evp_enc.c:357]: Doesn't a zero length IV lead to an undefined memcpy(3) a few lines later?

Comment (bernd-edlinger) [reply crypto/evp/evp_enc.c:357]: no only if iv == NULL at the same time.

Comment (paulidale) [reply crypto/evp/evp_enc.c:357]: Removed the equality test.

File: crypto/evp/evp_enc.c
Comment (mspncp) [inline crypto/evp/evp_enc.c:370]: Strange asymmetry: here a zero IV length is not allowed (and wasn't previously)?

File: crypto/evp/evp_enc.c
Comment (mspncp) [inline crypto/evp/evp_enc.c:374]: Strange asymmetry: here a zero IV length is not allowed (and wasn't previously)?

Comment (paulidale) [reply crypto/evp/evp_enc.c:374]: I believe this got mentioned before which is why I changed the previous check.

Comment (paulidale) [reply crypto/evp/evp_enc.c:374]: This code is, however, inside an `iv != NULL` clause so a zero length would be really strange.

Comment (mspncp) [reply crypto/evp/evp_enc.c:374]: Ah, I see!

**Review by mspncp — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to 3.0 and master.  Thanks for the great feedback.

**Pull request closed without merging**
